src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5790,6 +5790,215 @@ address generate_avx_ghash_processBlocks() {
+      const Register s = r11;
+      const Register a_d = r12; //r12d
+      const Register b_d = r8; //r8d
+      const Register end = r13;
+
+      const XMMRegister ya = xmm0;
+      const XMMRegister yb = xmm1;
+      const XMMRegister ydata0 = xmm2;
+      const XMMRegister ydata1 = xmm3;
+      const XMMRegister ysa = xmm4;
+      const XMMRegister ydata = ysa;
+      const XMMRegister ytmp0 = ydata0;
+      const XMMRegister ytmp1 = ydata1;
+      const XMMRegister ytmp2 = xmm5;
+      const XMMRegister xa = xmm0;
+      const XMMRegister xb = xmm1;
+      const XMMRegister xtmp0 = xmm2;
+      const XMMRegister xtmp1 = xmm3;
+      const XMMRegister xsa = xmm4;
+      const XMMRegister xtmp2 = xmm5;
+      const XMMRegister yshuf0 = xmm6;
+      const XMMRegister yshuf1 = xmm7;
+      assert_different_registers(init_d, c_rarg1, c_rarg2, data, size, s, a_d, b_d, end, rax);
+
+      Label SLOOP1, SLOOP1A, SKIP_LOOP_1A, FINISH, LT64, DO_FINAL, FINAL_LOOP, ZERO_SIZE, END;
+
+      BLOCK_COMMENT("Entry:");
+      __ enter(); // required for proper stackwalking of RuntimeStub frame
+
+      __ push(r12);
+      __ push(r13);
+      __ vmovdqu(yshuf0, ExternalAddress((address) StubRoutines::x86::_adler32_shuf0_table));
+      __ vmovdqu(yshuf1, ExternalAddress((address) StubRoutines::x86::_adler32_shuf1_table));
+      __ movptr(data, c_rarg1); //data
+      __ movl(size, c_rarg2); //length
+      __ movl(b_d, init_d); //adler
+      __ shrl(b_d, 16);
+      __ andl(init_d, 0xFFFF);
+      __ cmpl(size, 32);
+      __ jcc(Assembler::below, LT64);
+      __ movdl(xa, init_d); //vmovd - 32bit
+      __ vpxor(yb, yb, yb, Assembler::AVX_256bit);
+
+      __ BIND(SLOOP1);
+      __ movl(s, LIMIT);
+      __ cmpl(s, size);
+      __ cmovl(Assembler::above, s, size); // s = min(size, LIMIT)
+      __ lea(end, Address(s, data, Address::times_1, -CHUNKSIZE_M1));
+      __ cmpq(data, end);
+      __ jcc(Assembler::aboveEqual, SKIP_LOOP_1A);
+
+      __ BIND(SLOOP1A);
+      __ vbroadcastf128(ydata, Address(data, 0), Assembler::AVX_256bit);
+      __ addptr(data, CHUNKSIZE);
+      __ vpshufb(ydata0, ydata, yshuf0, Assembler::AVX_256bit);
+      __ vpaddd(ya, ya, ydata0, Assembler::AVX_256bit);
+      __ vpaddd(yb, yb, ya, Assembler::AVX_256bit);
+      __ vpshufb(ydata1, ydata, yshuf1, Assembler::AVX_256bit);
+      __ vpaddd(ya, ya, ydata1, Assembler::AVX_256bit);
+      __ vpaddd(yb, yb, ya, Assembler::AVX_256bit);
+      __ cmpptr(data, end);
+      __ jcc(Assembler::below, SLOOP1A);
+
+      __ BIND(SKIP_LOOP_1A);
+      __ addptr(end, CHUNKSIZE_M1);
+      __ testl(s, CHUNKSIZE_M1);
+      __ jcc(Assembler::notEqual, DO_FINAL);
+
+      // either we're done, or we just did LIMIT
+      __ subl(size, s);
+
+      // reduce
+      __ vpslld(yb, yb, 3, Assembler::AVX_256bit); //b is scaled by 8
+      __ vpmulld(ysa, ya, ExternalAddress((address) StubRoutines::x86::_adler32_ascale_table), Assembler::AVX_256bit); //need scratch register??
```

Comment:
All the instructions with ExternalAddress can modify rscratch1 which is r10. It is good to pass an explicit scratch register to these as last argument.